### DISCUSSION TITLE
Looking for input: Protobuf style binary format for encoding messages (CC1101)

### DISCUSSION THREAD

**Discussion by beardstack** [jgromes/RadioLib#673] [Ideas]
Hey there;

I am looking to create a standard format to store captured signals from URH and be able to edit, save and replay them on a CC1101 transceiver. I have too many signals to capture and there is no standardized format out there when it comes to this. I've mostly been working with the evilcrow hardware but would also eventually like to be able to use it elsewhere too (e.g yardstick one).

I'm looking to optimize space utilization, reduce processing (eliminate json) by using protobuf and add some extra metadata making things easier to deal with on a micro controller. At this present time I am focused on the CC1101 chip but ideally I'd really like a standard format for more than this particular transmitter.

I am both new to RadioLib, signal transmissions and protobuf. I'd really appreciate any comment on the following protobuf package/message structure. I want to use best practices, sensible defaults and use a sensible structure that contains every detail possible to allow replaying of any compatible signal. The next step after this would be to create a new plugin in URH to export this particular message format. What are your thoughts?

`[github](https://github.com/beardstack/urh/tree/master/protobuf)`

```
[syntax = "proto3";

package cc11011signal;

enum Modulation {
  MOD_ASK = 0;
  MOD_FSK = 1;
  MOD_PSK = 2;
}


enum EncodingFormat {
  ENCODING_NRZ = 0; //RADIOLIB_ENCODING_NRZ
  ENCODING_NRZ_INVERT = 1;
  ENCODING_MANCHESTER_1 = 2; //RADIOLIB_ENCODING_MANCHESTER
  ENCODING_MANCHESTER_2 = 3;
  ENCODING_DIFF_MANCHESTER = 4;
  ENCODING_WHITENING = 5; //RADIOLIB_ENCODING_WHITENING
}

message Message {
  optional int32 beginning_pause_ms = 1; //default 0
  optional int32 end_pause_ms = 2; //default 0
  repeated int32 data = 3;
  int32 length = 4; //length of bit array
}

message SyncWord {
  repeated int32 data = 1;
  int32 length = 2;
  int32 maxErrBits = 3;
}

message Node {
  repeated int32 Address = 1;
  numBroadcastAddrs = 2;
}


message Signal {

  required string name = 1; //label your message
  //CC1101 settings
  required float freq = 2;   //frequency for the message
  required float bitrate = 3; //bitrate,baud
  optional float deviation = 4;  //Not sure what that means, it's an option in evilcrow
  optional float rxBw = 5;
  optional int32 power = 6;
  
  //Message settings
  required Modulation modulation = 7; //One of the enums (ASK,FSK,PSK...)  
  repeated Message message = 8;   //The data array and pauses
  required EncodingFormat encoding = 9; //the encoding format
  
  //Extra optional fields
  optional bool enableCarrierSense = 10; //SyncWordFiltering
  optional bool enableOOK = 11; //OOK enable or disable
  optional bool enableCRCFiltering = 12; //CRC filtering enable or disable
  optional int32 DataShaping = 13; //???
  optional Node nodeAddr = 14; ///Destination info??
  
  optional string comment = 15; //User field](https://raw.githubusercontent.com/beardstack/urh/master/protobuf/cc1101_signal.proto)

}

```
EDIT: Updated after further tweaks

**Comment by beardstack** [comment]
hi @jgromes - may I have your opinion on this please?

**Reply by jgromes** [reply]
I would suggest to take a look at https://github.com/gnuradio/SigMF - it might not be exactly what you're looking for, but maybe a good inspiration?
